### PR TITLE
feat(apps): added endpoint to fetch app active documents

### PR DESCRIPTION
**Pull request by VPrasannaK94**
## Description

added endpoint to fetch app active documents

## Why

Added endpoint to fetch Active app documents

## Issue

N/A Ref : [CPLP-3087]

[Link to pull request from other repository.](https://github.com/eclipse-tractusx/portal-cd/pull/89)

## Checklist

Please delete options that are not relevant.

- [x] I have followed the [contributing guidelines](https://github.com/eclipse-tractusx/portal-assets/blob/main/developer/Technical%20Documentation/Dev%20Process/How%20to%20contribute.md#commit-and-pr-guidelines)
- [x] I have performed a self-review of my own code
- [x] I have successfully tested my changes locally
- [x] I have added tests that prove my changes work
- [x] I have checked that new and existing tests pass locally with my changes
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/portalbackend/PortalBackend.DBAccess/Repositories/OfferRepository.cs [hunk 1]
```diff
@@ -822,4 +822,19 @@ public void AttachAndModifyAppInstanceSetup(Guid appInstanceSetupId, Guid offerI
...
+        .Where(o => o.Id == offerId &&
+            o.OfferStatusId == OfferStatusId.ACTIVE &&
+            o.OfferTypeId == offerTypeId &&
+            o.ProviderCompanyId == userCompanyId)
+        .Select(o => o.Documents.Where(doc => documentTypeIds.Contains(doc.DocumentTypeId))
+            .Select(doc => new DocumentTypeData(
+                doc.DocumentTypeId,
+                doc.Id,
+                doc.DocumentName)))
+        .SingleOrDefaultAsync();
```
Comment (ntruchsess) [inline src/portalbackend/PortalBackend.DBAccess/Repositories/OfferRepository.cs:838]: returning Task<IEnumerable<...>?> from a query is an antipattern. You just don't do this, a query should allways return IAsyncEnumerable<...> instead.
I suggest:
```
_context.OfferAssignedDocuments
    .Where(x => x.OfferId == offerId &&
        x.Offer!.OfferStatusId == OfferStatusId.ACTIVE &&
        x.Offer.OfferTypeId == offerTypeId &&
        x.Offer.ProviderCompanyId == userCompanyId &&
        documentTypeIds.Contains(x.Document!.DocumentTypeId))
    .Select(x => new DocumentTypeData(
        x.Document!.DocumentTypeId,
        x.Document.Id,
        x.Document.DocumentName))
    .OrderBy(x => x.DocumentTypeId)
    .ToAsyncEnumerable();
```
(the OrderBy to be executed on the database is there so the businessCode may use AsyncGroupByExtensions.PreSortedGroupBy (namespace Org.Eclipse.TractusX.Portal.Backend.Framework.Async) to do the grouping afterwards.

File: src/marketplace/Apps.Service/BusinessLogic/AppChangeBusinessLogic.cs [hunk 2]
```diff
@@ -283,4 +283,17 @@ private async Task UpdateTenantUrlAsyncInternal(Guid offerId, Guid subscriptionI
...
         await _portalRepositories.SaveAsync().ConfigureAwait(false);
     }
+
+    /// <inheritdoc />
+    public async Task<ActiveAppDocumentData> GetActiveAppDocumentTypeDataAsync(Guid appId, Guid companyId)
+    {
+        var documentData = new Dictionary<DocumentTypeId, IEnumerable<DocumentData?>>();
+        var results = await _portalRepositories.GetInstance<IOfferRepository>().GetActiveOfferDocumentTypeDataAsync(appId, companyId, OfferTypeId.APP, _settings.ActiveAppDocumentTypeIds).ConfigureAwait(false);
+        var appDocTypeData = results!.GroupBy(d => d.DocumentTypeId).ToDictionary(g => g.Key, g => g.Select(d => new DocumentData(d.DocumentId, d.DocumentName)));
+        foreach (var doctype in _settings.ActiveAppDocumentTypeIds)
```
Comment (ntruchsess) [inline src/marketplace/Apps.Service/BusinessLogic/AppChangeBusinessLogic.cs:293]: the whole thing can be done efficiently using pure LINQ:
```
var appDocTypeData = await _portalRepositories.GetInstance<IOfferRepository>()
    .GetActiveOfferDocumentTypeDataAsync(appId, companyId, OfferTypeId.APP, _settings.ActiveAppDocumentTypeIds)
    .PreSortedGroupBy(result => result.DocumentTypeId)
    .ToDictionaryAsync(
        group => group.Key,
        group => group.Select(result =>
            new DocumentData(
                result.DocumentId,
                result.DocumentName)))
    .ConfigureAwait(false);
return new ActiveAppDocumentData(
    _settings.ActiveAppDocumentTypeIds.ToDictionary(
        documentTypeId => documentTypeId,
        documentTypeId => appDocTypeData.TryGetValue(documentTypeId, out var data)
            ? data
            : Enumerable.Empty<DocumentData>()));
```
(suggestion is based on the query returning IAsyncEnumerable as suggested in the comment in OfferRepository.cs)

File: src/marketplace/Apps.Service/ViewModels/AppData.cs [hunk 3]
```diff
@@ -43,3 +44,9 @@ public record AppData(
     string Price,
     Guid LeadPictureId,
     IEnumerable<string> UseCases);
+
+/// <summary>
+/// View model of an application's base data.
+/// </summary>
+/// <param name="Documents">Id of the App.</param>
+public record ActiveAppDocumentData(IDictionary<DocumentTypeId, IEnumerable<DocumentData?>> Documents);
```
Comment (ntruchsess) [inline src/marketplace/Apps.Service/ViewModels/AppData.cs:52]: DocumentData being nullable within the IEnumerable makes no sense. In the endpoints response you don't want to see 'null' as a member of the json-array that is mapped to the documentTypeId. The code that is creating those objects needs to be adjusted accordingly so it doesn't put null into this Enumerable.

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/marketplace/Apps.Service/Controllers/AppChangeController.cs [hunk 4]
```diff
@@ -186,4 +186,19 @@ public async Task<NoContentResult> UpdateTenantUrl([FromRoute] Guid appId, [From
...
+    /// <remarks>Example: GET /apps/appchange/{appId}/documents</remarks>
+    /// <response code="200">Gets the Active Apps documents</response>
+    /// <response code="404">If App does not exists</response>
+    [HttpGet]
+    [Route("{appId}/documents")]
+    [Authorize(Roles = "edit_apps")]
+    [ProducesResponseType(typeof(ActiveAppDocumentData), StatusCodes.Status200OK)]
+    [ProducesResponseType(typeof(ErrorResponse), StatusCodes.Status404NotFound)]
+    public async Task<ActiveAppDocumentData> GetActiveAppDocuments([FromRoute] Guid appId) =>
+        await this.WithCompanyId(companyId => _businessLogic.GetActiveAppDocumentTypeDataAsync(appId, companyId)).ConfigureAwait(false);
```
Comment (ntruchsess) [inline src/marketplace/Apps.Service/Controllers/AppChangeController.cs:203]: please use the new IIdentityService in the businessLogic to retrieve the companyId of the calling user.

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/marketplace/Apps.Service/BusinessLogic/AppChangeBusinessLogic.cs (hunk 2 above)
Comment (VPrasannaK94) [reply src/marketplace/Apps.Service/BusinessLogic/AppChangeBusinessLogic.cs:293]: @ntruchsess , In Ticket, it mentioned If for any of the types no document should be available, the document type should be still available in the response body, but the array should be empty.
So, we need to have all configured activeAppDocumentTypeIds in responce body.

@jjeroch, Can you please clarify it from your end also.

File: tests/portalbackend/PortalBackend.DBAccess.Tests/OfferRepositoryTests.cs [hunk 5]
```diff
@@ -1441,6 +1441,36 @@ public async Task GetServiceAccountProfileDataForSubscription_WithServiceAccount
...
+        var sut = await CreateSut().ConfigureAwait(false);
+
+        // Act
+        var result = await sut.GetActiveOfferDocumentTypeDataOrderedAsync(
+            new("ac1cf001-7fbc-1f2f-817f-bce0572c0007"),
+            new("2dc4249f-b5ca-4d42-bef1-7a7a950a4f87"),
+            OfferTypeId.APP, activeDocumentType).ToListAsync().ConfigureAwait(false);
+
+        // Assert
+        result.Should().NotBeNull().And.HaveCount(1).And.Satisfy(
```
Comment (ntruchsess) [inline tests/portalbackend/PortalBackend.DBAccess.Tests/OfferRepositoryTests.cs:1465]: please add more documents to test-seeding, covering at lease 3 documentTypeIds and assert the query does return the results ordered by documentTypeid (this is a prerequisite for the call to PreSortedGroupBy in the businesslogic)

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/marketplace/Apps.Service/BusinessLogic/AppChangeBusinessLogic.cs (hunk 2 above)
Comment (ntruchsess) [reply src/marketplace/Apps.Service/BusinessLogic/AppChangeBusinessLogic.cs]: just clarified with @jjeroch : documentTypeIds that do not have documents shall be included as empty arrayes in the response. No adjusments of the current implementation required

File: src/portalbackend/PortalBackend.DBAccess/Repositories/OfferRepository.cs (hunk 1 above)
Comment (ntruchsess) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/OfferRepository.cs]: I see this query would not translate to sql due to the position of the OrderBy.
I've changed that. Executing the OrderBy on the level of the original entity works fine. See: https://github.com/catenax-ng/tx-portal-backend/commit/59a3fb9d6cb25aa51c8519a086fa735ba63e26eb

File: tests/portalbackend/PortalBackend.DBAccess.Tests/OfferRepositoryTests.cs (hunk 5 above)
Comment (ntruchsess) [reply tests/portalbackend/PortalBackend.DBAccess.Tests/OfferRepositoryTests.cs]: did add more test-data to cover the 'BeInAscentingOrder(x => x.DocumentTypeId)' assert

**Review by ntruchsess — APPROVED** [review]

**Pull request merged**